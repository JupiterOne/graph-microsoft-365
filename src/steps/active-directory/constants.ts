// Step IDs
export const STEP_ACCOUNT = 'account';
export const STEP_GROUPS = 'groups';
export const STEP_GROUP_MEMBERS = 'group-members';
export const STEP_USERS = 'users';

// Graph objects
export const ACCOUNT_ENTITY_TYPE = 'microsoft_365_account';
export const ACCOUNT_ENTITY_CLASS = 'Account';

export const GROUP_ENTITY_TYPE = 'microsoft_365_user_group';
export const GROUP_ENTITY_CLASS = 'UserGroup';

export const USER_ENTITY_TYPE = 'microsoft_365_user';
export const USER_ENTITY_CLASS = 'User';

/**
 * The entity type used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_TYPE = 'microsoft_365_group_member';

/**
 * The entity class used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_CLASS = 'User';

export const ACCOUNT_GROUP_RELATIONSHIP_TYPE =
  'microsoft_365_account_has_group';
export const ACCOUNT_USER_RELATIONSHIP_TYPE = 'microsoft_365_account_has_user';
export const GROUP_MEMBER_RELATIONSHIP_TYPE = 'microsoft_365_group_has_member';
