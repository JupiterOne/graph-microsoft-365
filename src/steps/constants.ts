export const IngestionSources = {
  GROUPS: 'groups',
  USERS: 'users',
  APPLICATIONS: 'applications',
  DEVICES: 'devices',
  COMPLIANCE_POLICIES: 'compliance_policies',
  DEVICE_CONFIGURATIONS: 'device_configurations',
};
