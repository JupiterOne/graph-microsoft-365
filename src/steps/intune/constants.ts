import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
  FETCH_DEVICE_CONFIGURATIONS_AND_FINDINGS:
    'device-configurations-and-findings',
  FETCH_MANAGED_APPLICATIONS: 'managed-applications',
  FETCH_DETECTED_APPLICATIONS: 'detected-applications',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Device', 'Host'],
  },
  DEVICE_CONFIGURATION: {
    resourceName: 'Device Configuration',
    _type: 'intune_device_configuration',
    _class: 'Configuration',
  },
  NONCOMPLIANCE_FINDING: {
    resourceName: 'Noncompliance Finding',
    _type: 'intune_noncompliance_finding',
    _class: 'Finding',
  },
  MANAGED_APPLICATION: {
    resourceName: 'Managed Application',
    _type: 'intune_managed_application',
    _class: 'Application',
  },
  DETECTED_APPLICATION: {
    resourceName: 'Detected Application',
    _type: 'intune_detected_application',
    _class: 'Application',
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'azure_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
  DEVICE_USES_DEVICE_CONFIGURATION: {
    _type: 'intune_managed_device_uses_device_configuration',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.USES,
    targetType: entities.DEVICE_CONFIGURATION._type,
  },
  DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_device_configuration_identified_noncompliance_finding',
    sourceType: entities.DEVICE_CONFIGURATION._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  DEVICE_HAS_NONCOMPLIANCE_FINDING: {
    _type: 'intune_managed_device_has_noncompliance_finding',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  DEVICE_ASSIGNED_MANAGED_APPLICATION: {
    _type: 'intune_managed_device_assigned_application',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.MANAGED_APPLICATION._type,
  },
  DEVICE_HAS_DETECTED_APPLICATION: {
    _type: 'intune_managed_device_has_detected_application',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DETECTED_APPLICATION._type,
  },
};
