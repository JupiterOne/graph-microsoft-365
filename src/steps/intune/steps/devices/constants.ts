import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../../../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Host', 'Device'],
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'microsoft_365_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
};
