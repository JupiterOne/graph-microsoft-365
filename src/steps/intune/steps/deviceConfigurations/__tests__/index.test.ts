import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { setupAzureRecording } from '../../../../../../test/recording';
import { config } from '../../../../../../test/config';
import { fetchDeviceConfigurations, isOrphanedConfig } from '..';
import { entities } from '../../../constants';
import { DeviceConfiguration } from '@microsoft/microsoft-graph-types-beta';
import { DeviceConfigurationDeviceOverview } from '@microsoft/microsoft-graph-types';

describe('fetchDeviceConfigurations', () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) {
      await recording.stop();
    }
  });

  test('should make entities and relationships correctly', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchDeviceConfigurations',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });
    await fetchDeviceConfigurations(context);

    const deviceConfigurationEntities = context.jobState.collectedEntities;

    // Check that we have configurations
    expect(deviceConfigurationEntities.length).toBeGreaterThan(0);

    // Check that we have only ingested Device Configurationsd
    deviceConfigurationEntities.forEach((configuration) => {
      expect(configuration._type).toBe(entities.DEVICE_CONFIGURATION._type);
    });

    // Check that there are no orphaned configurations
    deviceConfigurationEntities.forEach((configuration) => {
      expect(
        isOrphanedConfig(
          ((configuration._rawData as unknown) as {
            rawData: DeviceConfiguration & {
              deviceStatusOverview: DeviceConfigurationDeviceOverview;
            };
          }[])[0].rawData.deviceStatusOverview,
        ),
      ).toBe(false);
    });

    // Check that the schema is correct
    expect(deviceConfigurationEntities).toMatchGraphObjectSchema({
      _class: entities.DEVICE_CONFIGURATION._class,
    });

    expect(deviceConfigurationEntities).toMatchSnapshot(
      'deviceConfigurationEntities',
    );
  });
});
