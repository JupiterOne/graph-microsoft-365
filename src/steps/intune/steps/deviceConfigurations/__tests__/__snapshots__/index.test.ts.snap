// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchDeviceConfigurations should make entities and relationships correctly: deviceConfigurationEntities 1`] = `
Array [
  Object {
    "_class": Array [
      "Configuration",
    ],
    "_key": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.iosCustomConfiguration",
          "createdDateTime": "2021-02-17T18:31:21.3225031Z",
          "description": null,
          "deviceStatusOverview": Object {
            "configurationVersion": 1,
            "errorCount": 1,
            "failedCount": 0,
            "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b",
            "lastUpdateDateTime": "2021-03-01T00:41:32.6166667Z",
            "notApplicableCount": 0,
            "pendingCount": 0,
            "successCount": 0,
          },
          "deviceStatusOverview@odata.context": "https://graph.microsoft.com/v1.0/$metadata#deviceManagement/deviceConfigurations('6aa605f0-744e-4ae1-8031-c6564624c89b')/microsoft.graph.iosCustomConfiguration/deviceStatusOverview/$entity",
          "displayName": "Test configuration policy",
          "id": "6aa605f0-744e-4ae1-8031-c6564624c89b",
          "lastModifiedDateTime": "2021-02-17T18:31:21.3225031Z",
          "payload": "MTk3",
          "payloadFileName": "lastSuccessfulBuildId.txt",
          "payloadName": "Test fake profile",
          "version": 1,
        },
      },
    ],
    "_type": "intune_device_configuration",
    "configurationType": "iosCustomConfiguration",
    "createdOn": 1613586681322,
    "description": "Test configuration policy",
    "displayName": "Test configuration policy",
    "id": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "name": "Test configuration policy",
    "version": 1,
  },
]
`;
